### PR TITLE
Updates to InlineAudio to support tts autoplay

### PR DESCRIPTION
**Pull request by maureensturgeon**
Adds a new prop, `ttsAutoplayEnabled` to the `InlineAudio` component. When true, the audio will automatically play when the component mounts. If the audio autoplay fails to play because of a DOMException, listeners will be added to the specified element if it exists (`props. autoplayTriggerElementId`) or document to play audio after user interaction.

These changes are broken out from the larger group of changes needed to enable automatic TTS autoplay instructions. The full set of changes can be found [here](https://github.com/code-dot-org/code-dot-org/pull/38956). 

## Testing story

These changes were tested in the larger PR https://github.com/code-dot-org/code-dot-org/pull/38956. I turned on autoplay for a teacher's section, logged in as a student on Chrome, went to a lesson, checked that instructions were autoplaying. On reload there is a DOMException, I checked that interactions on the workspace trigger the instructions to play once. 

I additionally verified on this branch that the InlineAudio still works with the new prop.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/instructions/InlineAudio.jsx [hunk 1]
```diff
@@ -90,6 +104,11 @@ class InlineAudio extends React.Component {
     src: PropTypes.string,
     message: PropTypes.string,
     style: PropTypes.object,
+    ttsAutoplayEnabled: PropTypes.bool,
```
Comment (maureensturgeon) [inline apps/src/templates/instructions/InlineAudio.jsx:107]: This is added, but not being set to true anywhere in the code yet

File: apps/test/unit/templates/instructions/InlineAudioTest.js [hunk 2]
```diff
@@ -104,16 +113,43 @@ describe('InlineAudio', function() {
...
+        assetUrl={function() {}}
+        ttsAutoplayEnabled={false}
+      />
+    );
+
+    expect(component.state().playing).to.be.false;
+    await component.instance().playAudio();
+    expect(component.state().playing).to.be.true;
+  });
+
```
Comment (maureensturgeon) [inline apps/test/unit/templates/instructions/InlineAudioTest.js:150]: I attempted for 2 days to figure out how to stub the DOMException and check that playing audio starts up after the first user interaction but was getting a lot of trouble with the promises involved in that code path and tests not waiting on promises. I involved other engineers and ultimately decided it was not worth the time it was taking. Please forgive this slight lack of coverage.

File: apps/src/templates/instructions/InlineAudio.jsx [hunk 3]
```diff
@@ -210,6 +237,55 @@ class InlineAudio extends React.Component {
...
 
+  getAutoplayTriggerElement() {
+    const {autoplayTriggerElementId} = this.props;
+    return autoplayTriggerElementId
+      ? document.getElementById(autoplayTriggerElementId)
+      : document;
+  }
+
+  // adds event listeners from the dom which trigger audio
+  // when a significant enough user has happened
```
Comment (cforkish) [inline apps/src/templates/instructions/InlineAudio.jsx]: typo

File: apps/src/templates/instructions/InlineAudio.jsx [hunk 4]
```diff
@@ -210,6 +237,55 @@ class InlineAudio extends React.Component {
     });
   }
 
+  getAutoplayTriggerElement() {
```
Comment (cforkish) [inline apps/src/templates/instructions/InlineAudio.jsx]: i'm not sure how expensive `getElementById` is, but since you're calling this for each enabling dom event, would it be better to set `this.listeningElement = getAutoplayTriggerElement()` in the constructor so you can just reference it?

File: apps/src/templates/instructions/InlineAudio.jsx [hunk 5]
```diff
@@ -210,6 +237,55 @@ class InlineAudio extends React.Component {
...
+    return autoplayTriggerElementId
+      ? document.getElementById(autoplayTriggerElementId)
+      : document;
+  }
+
+  // adds event listeners from the dom which trigger audio
+  // when a significant enough user has happened
+  addAudioAutoplayTrigger() {
+    AUDIO_ENABLING_DOM_EVENTS.forEach(event => {
+      const listeningElement = this.getAutoplayTriggerElement();
```
Comment (cforkish) [inline apps/src/templates/instructions/InlineAudio.jsx]: if you don't want to store this in the constructor, i'd at least store it outside of the `forEach` (same in `remove` below)

File: apps/src/templates/instructions/InlineAudio.jsx [hunk 6]
```diff
@@ -210,6 +237,55 @@ class InlineAudio extends React.Component {
...
+        // https://developers.google.com/web/updates/2017/09/autoplay-policy-changes
+        if (err instanceof DOMException && shouldAutoPlay) {
+          this.addAudioAutoplayTrigger();
+        } else {
+          throw err;
+        }
+      });
+  }
+
+  triggerAudio() {
```
Comment (cforkish) [inline apps/src/templates/instructions/InlineAudio.jsx]: NIT: imo `autoplayAudio` would be a more consistent name for this. the rest of the naming suggests we have an `autoplay` action that we potentially want to trigger by an event -- i.e. a trigger event causes an `autoplayAudio` action, not a `triggerAudio` action.

File: apps/test/unit/templates/instructions/InlineAudioTest.js [hunk 7]
```diff
@@ -14,7 +14,16 @@ const DEFAULT_PROPS = {
   style: {
     button: {},
     buttonImg: {}
-  }
+  },
+  ttsAutoplayEnabled: false
+};
+
+// when this function is called and awaited, it will not
+// execute until all promises queued up before it are resolved
```
Comment (cforkish) [inline apps/test/unit/templates/instructions/InlineAudioTest.js]: NIT: this grammer is kinda hard to parse

File: apps/test/unit/templates/instructions/InlineAudioTest.js [hunk 8]
```diff
@@ -104,16 +113,43 @@ describe('InlineAudio', function() {
...
     expect(component.state().playing).to.be.false;
     component.instance().toggleAudio();
+    await waitForPromises();
     expect(component.state().playing).to.be.true;
     component.instance().toggleAudio();
+    await waitForPromises();
     expect(component.state().playing).to.be.false;
   });
 
+  it('autoplays if autoplay of text-to-speech is enabled', async function() {
```
Comment (cforkish) [inline apps/test/unit/templates/instructions/InlineAudioTest.js:126]: TIL: i didn't realize we could use async functions here! cool!

**Review by cforkish — APPROVED** [review]
LGTM! a couple minor comments but nothing blocking.

File: apps/test/unit/templates/instructions/InlineAudioTest.js (hunk 7 above)
Comment (maureensturgeon) [reply apps/test/unit/templates/instructions/InlineAudioTest.js]: Ah yeah it was hard to explain, I'll try again.

File: apps/src/templates/instructions/InlineAudio.jsx (hunk 5 above)
Comment (maureensturgeon) [reply apps/src/templates/instructions/InlineAudio.jsx]: I assign it to `this` in `componentDidMount`

File: apps/src/templates/instructions/InlineAudio.jsx [hunk 9]
```diff
@@ -103,11 +122,25 @@ class InlineAudio extends React.Component {
...
+  constructor(props) {
+    super(props);
+    this.autoplayAudio = this.autoplayAudio.bind(this);
+  }
+
   componentDidMount() {
     this.getAudioElement();
+    if (this.props.ttsAutoplayEnabled && !this.state.autoplayed) {
+      const {autoplayTriggerElementId} = this.props;
+      this.autoplayTriggerElement = autoplayTriggerElementId
```
Comment (cforkish) [inline apps/src/templates/instructions/InlineAudio.jsx:139]: NIT: i'm personally of the opinion that we should declare all instance properties in or under the constructor (depending on whether or not they need access to `this`) even when they start out as null as a sort of "directory" for future devs, so i'd recommend adding `autoplayTriggerElement = null` around line 33

File: apps/src/templates/instructions/InlineAudio.jsx (hunk 5 above)
Comment (cforkish) [reply apps/src/templates/instructions/InlineAudio.jsx]: 👍

File: apps/src/templates/instructions/InlineAudio.jsx [hunk 10]
```diff
@@ -210,6 +243,49 @@ class InlineAudio extends React.Component {
...
+  removeAudioAutoplayTrigger() {
+    AUDIO_ENABLING_DOM_EVENTS.forEach(event => {
+      this.autoplayTriggerElement.removeEventListener(
+        event,
+        this.autoplayAudio
+      );
+    });
+  }
+
+  playAudio() {
```
Comment (jamescodeorg) [inline apps/src/templates/instructions/InlineAudio.jsx:263]: What happens if playAudio() is called more than once?  (I think there are some potential race conditions where this might happen?)
